python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -641,16 +680,32 @@ def build_sensor_context(
     check.opt_inst_param(instance, "instance", DagsterInstance)
     check.opt_str_param(cursor, "cursor")
     check.opt_str_param(repository_name, "repository_name")
-    return SensorEvaluationContext(
-        instance_ref=None,
-        last_completion_time=None,
-        last_run_key=None,
-        cursor=cursor,
-        repository_name=repository_name,
-        instance=instance,
-        repository_def=repository_def,
-        sensor_name=sensor_name,
+
+    from dagster._core.execution.build_resources import build_resources
+
+    required_resource_keys = None
+    if repository_def and sensor_name:
+        sensor_def = repository_def.get_sensor_def(sensor_name)
+        required_resource_keys = sensor_def.required_resource_keys
+
+    top_level_resources = repository_def.get_top_level_resources() if repository_def else {}
+    resources_to_build = (
+        {k: v for k, v in top_level_resources.items() if k in required_resource_keys}
+        if required_resource_keys
+        else top_level_resources
     )
+    with build_resources(resources_to_build) as resources:
+        return SensorEvaluationContext(
+            instance_ref=None,
+            last_completion_time=None,
+            last_run_key=None,
+            cursor=cursor,
+            repository_name=repository_name,
+            instance=instance,
+            repository_def=repository_def,
+            sensor_name=sensor_name,
+            resources=resources,
+        )
```

Comment:
at first look this is suspicious seems like `build_sensor_context` should itself be a context manager that is only closed after the tick evaluation is complete? If this is deliberate, at a minimum this is worthy of a comment

Reply:
👍  migrated to using a pattern similar to `op_context` to initialize resources